### PR TITLE
BLD: add a meson windows build

### PR DESCRIPTION
**Pull request by mattip**
Add a meson build for windows. Eventually, we should extend this to replace the azure windows builds. In addition to the new github workflow this:
- Tweaks the meson.md instructions and the conda environment
- Changes imports to import `_distributor_init` before the c-extensions
- ~Changes `tools/openblas_support.py` to generate the `_distributor_init` and use `os.add_dll_directory()` to enable loading OpenBLAS~ (reverted this since it does not work with conda)
- Uses the `openblas*.dll` packaged by `MacPython/openblas-libs` instead of repackaging the `openblas.a`. No mingw is needed for building and packaging NumPy. It is needed for f2py, until we find a better fortran compiler.
- Wrap `_fdopen` in an [invalid parameter handler](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/set-invalid-parameter-handler-set-thread-local-invalid-parameter-handler?view=msvc-170). Without it, linking to the wrong runtime can cause the process to silently exit. Now when `_fdopen` fails, a helpful error message will be emitted.

### CODE REVIEW

File: numpy/random/meson.build [hunk 1]
```diff
@@ -66,7 +66,7 @@ random_pyx_sources = [
       fs.copyfile('mtrand.pyx'),
       'src/distributions/distributions.c',
       'src/legacy/legacy-distributions.c'
-    ], ['-DNPY_RANDOM_LEGACY=1'], npymath_lib,
+    ], ['-DNP_RANDOM_LEGACY=1'], npymath_lib,
```
Comment (mattip) [inline numpy/random/meson.build:69]: This typo took wayyyy too long to find. The build did not crash locally, only on CI, and it took quite a bit of effort to get meson to emit pdb files that I could download and use locally.

File: numpy/meson.build [hunk 2]
```diff
@@ -29,6 +30,9 @@ if is_windows
     endif
   endif
 endif
+if is_msvc
+      add_project_arguments('-DNDEBUG', language: ['c', 'cpp'])
+endif
```
Comment (mattip) [inline numpy/meson.build:35]: While maybe not strictly needed, it is conventional to add this to a release build

File: numpy/random/meson.build (hunk 1 above)
Comment (charris) [reply numpy/random/meson.build:69]: Maybe we should define `NPY_RANDOM_LEGACY` as an alias. `NPY` looks like the expected prefix.

Comment (mattip) [reply numpy/random/meson.build:69]: xref the meson tracking issue #22546

**Comment by mattip** [comment]
xref the meson tracking issue https://github.com/numpy/numpy/issues/22546

**Comment by mattip** [comment]
Maybe we should be caching more. The CircleCI, debug, and "old gcc" builds all failed to download ubuntu packages

**Comment by ganesh-k13** [comment]
This windows bug is present in NumPy as well https://github.com/scipy/scipy/pull/17936, I'll raise a PR soon. I'm surprised this [line](https://github.com/numpy/numpy/pull/23165/files#diff-661a1b4adf5fef221e477b7f22915d949bb01a43176304896a17bab59efde46cR86) didn't break

File: numpy/random/meson.build (hunk 1 above)
Comment (mattip) [reply numpy/random/meson.build:69]: It is legacy so I don't think we should change/alias it. But it sure was annoying to find. I wonder if there is a way we could define it inside mtrand.pyx instead, but that is beyond this PR.

File: numpy/__init__.py [hunk 3]
```diff
@@ -112,6 +112,9 @@
     ComplexWarning, ModuleDeprecationWarning, VisibleDeprecationWarning,
     TooHardError, AxisError)
 
+# Allow distributors to run custom init code before importing numpy.core
+from . import _distributor_init
```
Comment (rgommers) [inline numpy/__init__.py:116]: This changes this import to happen at build-time rather than only at runtime. Distributors may have content here that relies on numpy being built (a use case for this file is modifying numpy defaults). So I think this should stay inside the `else:` clause. If there's a reason to move it, can you move it to right below `else:`?

File: numpy/core/include/numpy/npy_3kcompat.h [hunk 4]
```diff
@@ -167,6 +167,24 @@ static inline int PyInt_Check(PyObject *op) {
...
 
+/*
+ * Macros to protect CRT calls against instant termination when passed an
+ * invalid parameter (issue23524).
+ */
+#if defined _MSC_VER && _MSC_VER >= 1900
+
+extern _invalid_parameter_handler _Py_silent_invalid_parameter_handler;
+#define NPY_BEGIN_SUPPRESS_IPH { _invalid_parameter_handler _Py_old_handler = \
+    _set_thread_local_invalid_parameter_handler(_Py_silent_invalid_parameter_handler);
```
Comment (rgommers) [inline numpy/core/include/numpy/npy_3kcompat.h:178]: This looked a little fishy to me at first, but this function is documented by MSVC even though it's apparently private: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/set-invalid-parameter-handler-set-thread-local-invalid-parameter-handler?view=msvc-170.

One question: it's supposed to come from `stdlib.h`, so isn't it better to explicitly include that here? Like in CPython: https://github.com/python/cpython/blob/36854bbb240e417c0df6f0014924fcc899388186/Include/internal/pycore_fileutils.h#L264

File: numpy/meson.build (hunk 2 above)
Comment (rgommers) [reply numpy/meson.build:35]: But this is not for a release build, this is unconditional? There's some discussion at https://stackoverflow.com/questions/2290509/debug-vs-ndebug, and it may be related to gh-23122. Not sure what is right here, but a comment about why this is here would be great if we need to keep it.

File: numpy/random/meson.build (hunk 1 above)
Comment (rgommers) [reply numpy/random/meson.build:69]: ugh yes, this is ugly and that was an unfortunate typo.

Comment (rgommers) [reply numpy/random/meson.build:69]: removing it from the build definition files sound like a nice follow-up. I think I agree with not adding the `NPY_` flavor.

File: tools/openblas_support.py [hunk 5]
```diff
@@ -232,36 +232,13 @@ def make_init(dirname):
...
-                            DLL_filenames.append(filename)
-                    if len(DLL_filenames) > 1:
-                        import warnings
-                        warnings.warn("loaded more than 1 DLL from .libs:"
-                                      "\\n%s" % "\\n".join(DLL_filenames),
-                                      stacklevel=1)
+            import sys
+            extra_dll_dir = os.path.join(os.path.dirname(__file__), '.libs')
+            if sys.platform == 'win32' and os.path.isdir(extra_dll_dir):
+                os.add_dll_directory(extra_dll_dir)
```
Comment (rgommers) [inline tools/openblas_support.py:241]: This may not work for Python <3.10, see https://github.com/conda/conda/issues/10897#issuecomment-977576891. Is it actually needed? If not, I'd really prefer to not touch this thing.

**Review by rgommers — COMMENTED** [review]
Thanks @mattip! This looks pretty good. I have rebased, pushed one commit with minor changes, and added some questions for a few things that may be issues.

File: numpy/__init__.py (hunk 3 above)
Comment (mattip) [reply numpy/__init__.py:116]: The problem stems from `from numpy.__config__ import show` since on the meson build that newly imports `_multiarray_umath`, which requires the dll loading from `_distributor_init`. I can move the lines into the `else` clause.

File: numpy/core/include/numpy/npy_3kcompat.h (hunk 4 above)
Comment (mattip) [reply numpy/core/include/numpy/npy_3kcompat.h:178]: Line 15 has `#include <Python.h>` which has conditionally [`#include <stdlib.h>`](https://github.com/python/cpython/blob/36854bbb240e417c0df6f0014924fcc899388186/Include/Python.h#L23), which is why this works for me. It would be better to explicitly include it. I will add that.

File: numpy/meson.build (hunk 2 above)
Comment (mattip) [reply numpy/meson.build:35]: Maybe we could leave supporting a debug build for a future PR. I think meson should be adding that flag in general for release builds. As it is, it took quite an effort to work out the build command sequence
```
meson setup build --prefix=$PWD\build-install -Ddebug=false --optimization 2 --vsenv
meson compile -C build -v
cd build
meson install --no-rebuild
cd ..
```

I will add a comment

File: tools/openblas_support.py (hunk 5 above)
Comment (mattip) [reply tools/openblas_support.py:241]: FWIW, I think the problem is the opposite: PATH manipulation no longer works in pythons provided by python.org , but still works in conda-provided pythons up to python3.10. `add_dll_directory` did not exist (the code preceeds python3.8), so at the time preloading the dlls was the only way to make this support all the versions of python supported by NumPy. Since we dropped support for <3.7, all versions should have `add_dll_directory`, so this is much cleaner.

I will move it to a separate PR.

File: .github/workflows/linux_meson.yml [hunk 6]
```diff
@@ -26,7 +22,7 @@ permissions:
 
 jobs:
   meson_devpy:
-    if: "github.repository == 'numpy/numpy' && !contains(github.event.head_commit.message, '[ci skip]') && !contains(github.event.head_commit.message, '[skip ci]') && !contains(github.event.head_commit.message, '[skip github]')"
+    if: "github.repository == 'numpy/numpy'"
```
Comment (mattip) [inline .github/workflows/linux_meson.yml:25]: @rgommers: why did you remove the conditions from this stanza? Is there a replacement mechanism?

File: numpy/meson.build (hunk 2 above)
Comment (mattip) [reply numpy/meson.build]: Hmm, removing the define, now the [build fails](https://github.com/numpy/numpy/actions/runs/4230989854/jobs/7348965062#step:7:299) at an assert
```
assert(ts->field_buffer_length >= 2*(ssize_t)sizeof(Py_UCS4));
```

since `ssize_t` is not defined for windows at that point ...

File: tools/openblas_support.py (hunk 5 above)
Comment (rgommers) [reply tools/openblas_support.py]: > PATH manipulation no longer works in pythons provided by python.org

I don't think this is true. The current contents work fine, AFAIK we have not had any trouble with them for a couple of years now. Unless I missed the relevant issue? If there is no issue, I am pretty sure we should not touch this at all until Python 3.10 is the minimum supported version.

File: .github/workflows/linux_meson.yml (hunk 6 above)
Comment (rgommers) [reply .github/workflows/linux_meson.yml:25]: This has had builtin support for quite a while now, see https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs. So we shouldn't continue replication this boilerplate. I can open a separate PR to remove it from the other jobs now. Sound good?

File: tools/openblas_support.py (hunk 5 above)
Comment (rgommers) [reply tools/openblas_support.py]: Ah I see you already opened the other PR.

File: .github/workflows/linux_meson.yml (hunk 6 above)
Comment (mattip) [reply .github/workflows/linux_meson.yml:25]: Ahh, cool. It makes sense to remove this from the other github workflows then.

File: numpy/meson.build (hunk 2 above)
Comment (mattip) [reply numpy/meson.build]: fixed. Compilation without NDEBUG succeeds on windows.

**Comment by mattip** [comment]
The meson windows CI is passing, I think this is ready for another round of review.

**Review by rgommers — COMMENTED** [review]
This is looking quite good now. One issue remaining that should be addressed before merging: there's actually a test failure (see below), so the error propagation doesn't work correctly in the `test` step of the CI job.

Failure:
```
_________________________ test_array_api_entry_point __________________________

    @pytest.mark.xfail(
        sysconfig.get_config_var("Py_DEBUG") is not None,
        reason=(
            "NumPy possibly built with `USE_DEBUG=True ./tools/travis-test.sh`, "
            "which does not expose the `array_api` entry point. "
            "See https://github.com/numpy/numpy/pull/19800"
        ),
    )
    def test_array_api_entry_point():
        """
        Entry point for Array API implementation can be found with importlib and
        returns the numpy.array_api namespace.
        """
        eps = importlib.metadata.entry_points()
        try:
            xp_eps = eps.select(group="array_api")
        except AttributeError:
            # The select interface for entry_points was introduced in py3.10,
            # deprecating its dict interface. We fallback to dict keys for finding
            # Array API entry points so that running this test in <=3.9 will
            # still work - see https://github.com/numpy/numpy/pull/19800.
            xp_eps = eps.get("array_api", [])
>       assert len(xp_eps) > 0, "No entry points for 'array_api' found"
E       AssertionError: No entry points for 'array_api' found
E       assert 0 > 0
E        +  where 0 = len([])

eps        = {'console_scripts': [EntryPoint(name='normalizer', value='charset_normalizer.cli.normalizer:cli_detect', group='consol..._finalize', value='setuptools.dist:_Distribution.finalize_options', group='setuptools.finalize_distribution_options')]}
xp_eps     = []

..\build-install\Lib\site-packages\numpy\tests\test_public_api.py:495: AssertionError
=========================== short test summary info ===========================
FAILED ..\build-install\Lib\site-packages\numpy\tests\test_public_api.py::test_array_api_entry_point
= 1 failed, 29062 passed, 1112 skipped, 1308 deselected, 33 xfailed in 430.09s (0:07:10) =
```

And now that there's a pretty clean build log, a future improvement (not in this PR) may be to address these warnings:
```
..\numpy\core\src\umath\loops.c.src(494): warning C4244: '=': conversion from 'double' to 'npy_byte', possible loss of data

..\numpy\core\src\umath\loops_unary.dispatch.c.src(440): warning C4146: unary minus operator applied to unsigned type, result still unsigned

..\numpy\core\src\umath\scalarmath.c.src(525): warning C4146: unary minus operator applied to unsigned type, result still unsigned
 
..\numpy\core\src\umath\scalarmath.c.src(1793): warning C4068: unknown pragma 'GCC'
 
../numpy/core/src/multiarray/item_selection.c(605): warning C4098: 'npy_fastputmask': 'void' function returning a value
 
../numpy/core/src/npysort/radixsort.cpp(29): warning C4146: unary minus operator applied to unsigned type, result still unsigned
../numpy/core/src/npysort/radixsort.cpp(106): note: see reference to function template instantiation 'UT KEY_OF<T,UT>(UT)' being compiled
        with
        [
            UT=UT,
            T=npy_int
        ]
        
../numpy/fft/_pocketfft.c(80): warning C4244: 'function': conversion from 'size_t' to 'double', possible loss of data
```

**Comment by mattip** [comment]
So there are two problems:
- the failing test did not fail the CI run
- something is off with `importlib.metadata.entry_points().select(group="array_api")

**Comment by mattip** [comment]
I think I am using exactly the same stanza as the [workflow for scipy](https://github.com/scipy/scipy/blob/cac5d9aa8cc1ae3d27748a60c4da2972982430cc/.github/workflows/windows.yml#L86). Do errors there fail the build?

**Comment by rgommers** [comment]
> I think I am using exactly the same stanza as the [workflow for scipy](https://github.com/scipy/scipy/blob/cac5d9aa8cc1ae3d27748a60c4da2972982430cc/.github/workflows/windows.yml#L86). Do errors there fail the build?

They do AFAIK. Either way, don't want to investigate now. Maybe just switch to `./dev.py test` like in the Linux build?

> something is off with `importlib.metadata.entry_points().select(group="array_api")

The `[project.entry-points]` table in `pyproject.toml` is commented out (it cannot be enabled yet, it will interfere with the `setup.py` builds). Hence the entry point does not exist. I suggest skipping the test for now, maybe adding a comment to unskip it in the commented-out `pyproject.toml` content.

File: README.md [hunk 7]
```diff
@@ -37,7 +37,7 @@ Testing:
 
 NumPy requires `pytest` and `hypothesis`.  Tests can then be run after installation with:
 
-    python -c 'import numpy; numpy.test()'
+    python -c "import numpy, sys; sys.exit(numpy.test() is False)"
```
Comment (mattip) [inline README.md:40]: Fixing this so the exit code will indicate passing/failure

**Comment by mattip** [comment]
The test check is `sysconfig.get_config_var("Py_DEBUG") is not None`. This means it is xfailed on linux, since `sysconfig.get_config_value("Py_DEBUG")` is `"0"`, where on windows it is `None`.

**Comment by mattip** [comment]
I reverted my "smarter" change to check `Py_DEBUG` to what it was before.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks @mattip!

**Comment by mattip** [comment]
Whohoo!
